### ISSUE TITLE
Master Midi Channel

### ISSUE THREAD

**Issue by mheidt** [le1ca/alesisvsysex#5]
Would you think, that it is possible to have a python script, that changes the channel for a partial set?
In my usecase I would like to let the python host (zynthian.org) send SYSEX commands to the V25 so that the keys, wheels and knobs would listen to  a new channel.

Is it performing fast enough while playing?
I have the vision, that the buttons send a Program Change (AH...that is missing as option in your tool, btw) to the host (zynthian), that is triggering a python script which changes those midi channels.
But I wouldn't do it if the script would take more than a second to perform.

**Comment by abridgewater** [comment]
As far as reconfiguring the keyboard under script control goes, for at least the V25 the windows utility save file format (and the format for my fork of alesisvsysex) is a "raw" sysex message containing the configuration as sent to the controller. The configuration needs to be sent to the "second" MIDI channel (it's probably marked as being for "Editor" in the USB descriptors, but this doesn't seem to be reflected anywhere else), and a test can probably be done via something as simple as "cat savefile.v25 > /dev/snd/midiC2D1" (these paths likely to vary).  This is unlikely to work for a VI-series controller and will not work with the save files for the windows VMini editor.

**Comment by tim-littlefair** [comment]
I realize that this is a very old issue and that it is some time since it was raised and even longer since @tmick0 or the other developers (notably @abridgewater and @Baggypants) who have forked and extended this project have done any work, but I had a similar interest in relation to the Alesis V61 (original, not mark II) and I have created a further fork and started work on making it possible and made some progress (which I'll describe after making some observations on @mheidt's question.

In relation to whether program change can be sent to the V25, I don't have access to V25 hardware, but I've checked [the manual](https://www.alesis.com/rscdn/1074/documents/V25%20-%20User%20Guide%20-%20v1.2.pdf) and this advises that both the buttons and the pads on the V25 device can be configured to send program change messages (as they can on my V61).

My V61 work is happening on on [this branch](https://github.com/tim-littlefair/alesisvsysex/tree/v61-support), which was forked from @tmick0's repository but I have merged @Baggypants' and @abridgewater's work in before starting to add my own changes.  On top of this basis, with relatively little effort (<20 hours) I've managed to create a new model subclass for the V61 in which the Program Change option on a button is accepted when a .v61 SysEx file exported from the first party V61 Editor on Windows is loaded into alesisvsysex.  I can also load alesisvsysex from my hardware device.  

The work on my branch isn't yet ready to call releasable, the most significant problem being that loading from either a file or hardware causes a new editor window to be opened in front of the one which already exists, but if I can fix at least that one concern I'd be keen to do a pull request to @tmick0's repository if he is still watching it and happy to receive, or to the forks of @abridgewater or @Baggypants if either of them is keen to curate onward development, but @tmick0 prefers not to.  Ideally I think @tmick0's repo at the root of the fork tree would be the most discoverable location for potential users, but its up to him and the other prior contributors to decide whether they want to continue to be involved.  

There are some architectural and naming changes which I'd be keen to think about, mostly in the model/component area - but I'd like to discuss my ideas with the prior contributors who added extensions to support for additional models before amending their excellent existing work, so it would be helpful if we could work out who still wants to be involved and which repo should receive future PR's/issues.

**Comment by tim-littlefair** [comment]
I think I've got a baseline which would equally support the V25, V49 or V61 at [this commit](https://github.com/tim-littlefair/alesisvsysex/commit/46886e234ef03d648ac540e8816f55c61eac8068) on my fork, and for all of those devices would enable the pad and button (aka switch) controls to be set to Program Change mode to address this issue.

Note that only the V61 is tested by me (because I don't have a V25 or a V49), and for the moment I've not attempted to bring in the VMini or VI49 support implemented on branches because these would require significantly more change than extending along the original V-series (all of which use the same SysEx format and can be covered by a single model.py class).

I've written up what I've done so far in ridiculous detail [here](https://github.com/tim-littlefair/alesisvsysex/blob/82b9fe7af24284be7e274f5b0d2f7a9fd16ed75f/tl-fork-docs/fork-notes.md).  

The document above also contains proposals for extra work I could look at in future if there is interest in the community, including re-integrating the VMini and VI49 and other VI series devices, and also adding support for the original AKAI Professional MPK mini (which I own an instance of).  Note that I am prepared to attempt to support devices I don't have access to, but only if I have identified people who have the hardware and are are willing to test on my behalf.

I'd be happy to raise a PR on @tmick0's repository for the current V-series work if he is interested in testing it on the V25 he hopefully still owns.

Thanks again to @tmick0 and other contributors for work done to date on this interesting project.

**Comment by tmick0** [comment]
Hi @tim-littlefair, I'd be happy to review a PR though I no longer have the V25 to be able to test it out myself. I really appreciate all the work you, @abridgewater, and @Baggypants have done and I sincerely apologize for neglecting this project for so long. Cheers!

**Comment by tim-littlefair** [comment]
Hi @tmick0 - no apology necessary for stopping work on the project when it met your needs - especially as you no longer use/own the hardware so this isn't really an itch you benefit from scratching any more.

I will raise a PR for you to review and would value your comments on this set of changes - but it seems silly to burden you with code reviews on future changes if you don't have the hardware to verify that your original supported functionality for V25 isn't broken.  

I think that before attempting to merge @abridgewater's work I am going to need to make some architectural changes so that full elaboration of the Qt GUI can be deferred until after a keyboard is detected, so that different model classes (and serializers) can be bound to the GUI tabs according to whether the detected device is V-series, VMini or VI-series  rather than requiring a default model matching the V-series to be unconditionally created during startup.

Thanks again for starting this off - I originally came to it seeking understanding of how preset load/save worked for another rtmidi-dependent project I am working on, it has given me a platform for experimentation which has enabled me to answer all the questions I started with.

**Comment by abridgewater** [comment]
> I'd be keen to do a pull request to [@tmick0](https://github.com/tmick0)'s repository if he is still watching it and happy to receive, or to the forks of [@abridgewater](https://github.com/abridgewater) or [@Baggypants](https://github.com/Baggypants) if either of them is keen to curate onward development, but [@tmick0](https://github.com/tmick0) prefers not to.

While I still have my hardware (though my VI49 suffered a bit of an accident and I haven't been able to source replacement parts to fix it properly), between Python, Qt, and the bit with SNDRV_RAWMIDI_IOCTL_DRAIN, I'm not interested in doing much with this going forwards.

That said, I'm willing to answer questions about the changes I've made. I'm sure that there are _some_ things that aren't obvious, aren't in the comments, and aren't in the commit messages.

**Comment by tmick0** [comment]
PC support added in d27e77fe5915c763c4a04f15f7514a5498c02145. Thanks @tim-littlefair for the contribution.
